helper/schema/schema.go

```diff
@@ -1431,6 +1465,59 @@ func (m schemaMap) diffString(
 	return nil
 }
 
+// handleDiffSuppressOnRefresh visits each of the attributes set in "new" and,
+// if the corresponding schema sets both DiffSuppressFunc and
+// DiffSuppressOnRefresh, checks whether the new value is materially different
+// than the old and if not it overwrites the new value with the old one,
+// in-place.
+func (m schemaMap) handleDiffSuppressOnRefresh(oldState, newState *terraform.InstanceState) {
+	if newState == nil || oldState == nil {
+		return // nothing to do, then
+	}
+
+	// We'll populate this in the loop below only if we find at least one
+	// attribute which needs this analysis.
+	var d *ResourceData
+
+	oldAttrs := oldState.Attributes
+	newAttrs := newState.Attributes
+	for k, newV := range newAttrs {
+		oldV, ok := oldAttrs[k]
+		if !ok {
+			continue // no old value to compare with
+		}
+		if newV == oldV {
+			continue // no change to test
+		}
+
+		schemaList := addrToSchema(strings.Split(k, "."), m)
+		if len(schemaList) == 0 {
+			continue // no schema? weird, but not our responsibility to handle
+		}
+		schema := schemaList[len(schemaList)-1]
+		if !schema.DiffSuppressOnRefresh || schema.DiffSuppressFunc == nil {
+			continue // not relevant
+		}
+
+		if d == nil {
+			// We populate "d" only on demand, to avoid the cost for most
+			// existing schemas where DiffSuppressAlways won't be set.
+			var err error
+			d, err = m.Data(newState, nil)
+			if err != nil {
+				// Should not happen if we got far enough to be doing this
+				// analysis, but if it does then we'll bail out.
+				log.Printf("[WARN] schemaMap.handleDiffSuppressAlways failed to construct ResourceData: %s", err)
```

Comment:
```suggestion
				log.Printf("[WARN] schemaMap.handleDiffSuppressOnRefresh failed to construct ResourceData: %s", err)
```

Not for this pull request, per-se, but it would also be good if this method accepted `context.Context` so we can switch this to terraform-plugin-log so it gains all the relevant structured logging keys including the RPC, resource type, etc.